
#load libraries and data
```{r}
#load packages
library(tidyverse)
library(openxlsx)
library(ggtext)
library(Biostrings)
library(sangerseqR)
library(reshape2)
library(svglite)
library(exactRankTests)

#set paths
input_dir= "./input/"
output_dir= "./output/"

#read metadata
sites=read.xlsx(paste0(input_dir, "SangerAnalysisInput.xlsx"), sheet="gt_sites")
dna_samples = read.xlsx(paste0(input_dir, "SangerAnalysisInput.xlsx"), sheet="sangerseqno_dnasample")
reftable=read.xlsx(paste0(input_dir, "SangerAnalysisInput.xlsx"), sheet="reference")
primerinfo=read.xlsx(paste0(input_dir, "SangerAnalysisInput.xlsx"), sheet="primers")
parentinfo=read.xlsx(paste0(input_dir, "SangerAnalysisInput.xlsx"), sheet="dnasample_parents")
DSBinfo=read.xlsx(paste0(input_dir, "SangerAnalysisInput.xlsx"), sheet="DSB")
filterproblem=read.xlsx(paste0(input_dir, "SangerAnalysisInput.xlsx"), sheet="filter")
dna_orient = left_join(dna_samples, primerinfo, by="PrimerName") %>% select(SangerSeqNo, Orientation)
```

```{r}
#set options

#set to TRUE for more stricter calling of genotypes
strict=TRUE

#ratio of GT to wt peak intensity
GT_peak_ratio = 0.25

#same, but for good/bad zone calling
GT_peak_ratio_zones = 0.1

#length of sequence to match to find the GT sites
GT_fdr_len = 20

#length of sequence to find whether site adjacent areas are mutated
search_width = 40

#drawing options
theme.size=7 # in points
line.thick=0.469 # in points (although 0.469 pt here is equal to 1 pt in inkscape)
geom.text.size = theme.size / (1/0.3514598025)
windowsFonts(Verdana = windowsFont("Verdana"))
jsiglevel1=1
jsiglevel2=0.05
jsiglevel3=0.01
jsiglevel4=0.001

fill_colour_df=data.frame(value=factor(c("wt/wt", "wt/mut", "GT/wt", "GT/mut", "GT/GT", "mut/mut", "undetermined")), 
                        fill_colour=factor(c("cornflowerblue", "darkslategray3", "plum", "lightsalmon2", "indianred3", "lightgoldenrod2", "grey")),
                        fill_colour_order=factor(c(1, 2, 3, 4, 5, 6, 7)))
#make this
shape_df=data.frame(shape=c(19, 1, NA),
                    label=c("TRUE", "FALSE", "undetermined"))



```

```{r}
#########################################################################################
#find wt or GT bases in primary or secondary sequences
############################################################################################


#acquire the genome reference sequence
refsequence=DNAString(reftable$Sequence[1])
refsequence_RC = reverseComplement(refsequence)
ref_len=nchar(as.character(refsequence))
DSB_position = as.integer(DSBinfo[[1,1]])
FLANK_A_REF = substr(refsequence, 1, DSB_position)
FLANK_A_REF_100 = substr(FLANK_A_REF, nchar(FLANK_A_REF)-100, nchar(FLANK_A_REF))
FLANK_A_REF_search = substr(FLANK_A_REF_100, 1, 20)
FLANK_B_REF = substr(refsequence, DSB_position+1, ref_len)
FLANK_B_REF_100 = substr(FLANK_B_REF, 1, 100)
FLANK_B_REF_search = substr(FLANK_B_REF_100, nchar(FLANK_B_REF_100)-20, nchar(FLANK_B_REF_100))

#make empty dataframes
Total_data = tibble(Name = NULL)
Total_data_before = tibble(Name = NULL)
Total_data_after = tibble(Name = NULL)

files=list.files(path=input_dir, pattern=".ab1$")

#check whether the list of samples matches the list of the actual files
for (j in files){
  clean_ab1_name = str_replace(j, "_premix.ab1", "")
clean_ab1_name_2 = str_replace(clean_ab1_name, ".ab1", "")
 if (clean_ab1_name_2 %in% dna_samples$SangerSeqNo){
   next
 } else{
   stop(paste0("Filename ", clean_ab1_name_2, " not found in SangerAnalysisInput - sangerseqno_dnasample"))
 }
}

for (j in files){

#get the sanger sequencing data
abif1 = readsangerseq(paste0(input_dir, j))  

#get clean ab1 name
clean_ab1_name = str_replace(j, "_premix.ab1", "")
clean_ab1_name_2 = str_replace(clean_ab1_name, ".ab1", "")

#get the primer orientation info
current_primer=as.character(dna_samples %>% filter(SangerSeqNo == clean_ab1_name_2) %>% select(PrimerName) )
current_orientation=as.character(primerinfo %>% filter(PrimerName == current_primer) %>% select(Orientation))
  
#make empty dataframes
Subtotal_data = tibble(Name = NULL)
Subtotal_data_before = tibble(Name = NULL)
Subtotal_data_after = tibble(Name = NULL)
Subtotal_data_names = tibble(Name = NULL)

#get basecalls from the ab1 file, split in primary and secondary
abifbasecalls = makeBaseCalls(abif1,ratio=GT_peak_ratio)
  
#determine what allele is the same as the ref and extract the primary and secondary sequences
if (current_orientation == "FW"){
hetseqalleles = try(setAllelePhase(abifbasecalls,refsequence,trim5=50,trim3=300), silent=TRUE)
}else{
hetseqalleles = try(setAllelePhase(abifbasecalls,refsequence_RC,trim5=50,trim3=300), silent=TRUE)
}

#if sequence is really bad, program should skip it and move on
if (inherits(hetseqalleles,'try-error') == TRUE ){
  
  for (i in row.names(sites)){
  Site_name = as.character(sites %>% filter(row.names(sites) %in% i) %>% select(Name))
  Site_start = as.integer(sites %>% filter(row.names(sites) %in% i) %>% select(Location))
  GT_Seq =  as.character(sites %>% filter(row.names(sites) %in% i) %>% select(GT))
  Site_width = nchar(GT_Seq)
  Insertion = as.integer(sites %>% filter(row.names(sites) %in% i) %>% select(Insertion))
  
  #Get the wt seq from the ref (only for the site/region of interest)
  if (current_orientation == "FW"){
  WT_Seq = as.character(subseq(refsequence, start=Site_start, width=Site_width))
  }else{
    if (Insertion > 0){ 
      WT_Seq = as.character(subseq(refsequence, start=(Site_start-Insertion), width=Site_width))
    }else{
      WT_Seq = as.character(subseq(refsequence, start=Site_start, width=Site_width))
    }
  }
    currentData = data.frame(Name=Site_name,
                            Seq1="not found",
                            Seq2="not found",
                            WTseq=WT_Seq,
                            GTseq=GT_Seq,
                            SangerSeqNo=clean_ab1_name_2,
                            Prim_Seq = "",
                            Sec_Seq = "")
    
    Subtotal_data_names = rbind(Subtotal_data_names, currentData)

    
  }

}else{

PrimSeq = DNAString(hetseqalleles@primarySeq)
SecSeq = DNAString(hetseqalleles@secondarySeq)

for (i in row.names(sites)){
  #get info from table
  Site_name = as.character(sites %>% filter(row.names(sites) %in% i) %>% select(Name))
  Site_start = as.integer(sites %>% filter(row.names(sites) %in% i) %>% select(Location))
  GT_Seq =  as.character(sites %>% filter(row.names(sites) %in% i) %>% select(GT))
  Site_width = nchar(GT_Seq)
  Insertion = as.integer(sites %>% filter(row.names(sites) %in% i) %>% select(Insertion))
  
  #Get the wt seq from the ref (only for the site/region of interest)
  if (current_orientation == "FW"){
  WT_Seq = as.character(subseq(refsequence, start=Site_start, width=Site_width))
  }else{
    if (Insertion > 0){ 
      WT_Seq = as.character(subseq(refsequence, start=(Site_start-Insertion), width=Site_width))
    }else{
      WT_Seq = as.character(subseq(refsequence, start=Site_start, width=Site_width))
    }
  }

  #get the sequence preceding and following the site of interest to match
  if (current_orientation == "FW"){
  SeqToSearch=subseq(refsequence, start=Site_start-GT_fdr_len, width=GT_fdr_len)
  SeqToSearch_before=subseq(refsequence, start=Site_start-search_width, width=search_width)
  SeqToSearch_after=subseq(refsequence, start=Site_start+Site_width, width=search_width)
  }else{
  #search on the other side of the site of interest when using reverse sequencing primer
  if (Insertion > 0){
  SeqToSearch=subseq(refsequence_RC, start=(ref_len-((Site_start-1)+(Site_width-(1+Insertion))+GT_fdr_len)), width=GT_fdr_len)
  SeqToSearch_before=subseq(refsequence_RC, start=(ref_len-(Site_start-2)), width=search_width)
  SeqToSearch_after=subseq(refsequence_RC, start=(ref_len-((Site_start-1)+(Site_width-(1+Insertion))+search_width)), width=search_width)
  
  }else{
  SeqToSearch=subseq(refsequence_RC, start=(ref_len-((Site_start-1)+(Site_width-1)+GT_fdr_len)), width=GT_fdr_len)
  SeqToSearch_before=subseq(refsequence_RC, start=(ref_len-(Site_start-2)), width=search_width)
  SeqToSearch_after=subseq(refsequence_RC, start=(ref_len-((Site_start-1)+(Site_width-1)+search_width)), width=search_width)
  
  }
  }
  
  #find the SeqToSearch in the primary and secondary calls
  MatchPrim= matchPattern(SeqToSearch, PrimSeq, max.mismatch = 1)
  MatchSec= matchPattern(SeqToSearch, SecSeq, max.mismatch = 1)
  PositionAfterStartPrim = MatchPrim@ranges@start+(MatchPrim@ranges@width)
  PositionAfterStartSec = MatchSec@ranges@start+(MatchSec@ranges@width)
  
  
  
  #get the sequences of the site of interest
  #strict mode that looks for the sites in the primary and secondary sequences
  if (strict==TRUE){
  if(length(PositionAfterStartPrim) == 1){
    if (current_orientation == "FW"){
    Allele1Seq = as.character(subseq(PrimSeq, start=PositionAfterStartPrim, width=Site_width ))
    }else{
    Allele1Seq = as.character(reverseComplement(subseq(PrimSeq, start=PositionAfterStartPrim, width=Site_width )))
    }
  }else if(length(PositionAfterStartPrim) == 0){
    Allele1Seq = "not found"
  }else{
    Allele1Seq = "ambiguous"
  }
  
    if(length(PositionAfterStartSec) == 1){
    if (current_orientation == "FW"){
      Allele2Seq = as.character(subseq(SecSeq, start=PositionAfterStartSec, width=Site_width ))
    }else{
      Allele2Seq = as.character(reverseComplement(subseq(SecSeq, start=PositionAfterStartSec, width=Site_width )))}
  }else if(length(PositionAfterStartSec) == 0){
    Allele2Seq = "not found"
  }else{
    Allele2Seq = "ambiguous"
  }}
  #a lenient mode that only looks for the sites in the primary sequence
  if (strict==FALSE){
    if(length(PositionAfterStartPrim) == 1){
    if (current_orientation == "FW"){
    Allele1Seq = as.character(subseq(PrimSeq, start=PositionAfterStartPrim, width=Site_width ))
    Allele2Seq = as.character(subseq(SecSeq, start=PositionAfterStartPrim, width=Site_width ))
    }else{
    Allele1Seq = as.character(reverseComplement(subseq(PrimSeq, start=PositionAfterStartPrim, width=Site_width )))
    Allele2Seq = as.character(reverseComplement(subseq(SecSeq, start=PositionAfterStartPrim, width=Site_width )))
    }
  }else if(length(PositionAfterStartPrim) == 0){
    Allele1Seq = "not found"
    Allele2Seq = "not found"
  }else{
    Allele1Seq = "ambiguous"
    Allele2Seq = "ambiguous"
  }
  }
  
  #write it to an object
  currentData = tibble(Name=Site_name, Seq1=Allele1Seq, Seq2=Allele2Seq) %>% 
    mutate(WTseq = WT_Seq, 
           GTseq = GT_Seq)
  
  Subtotal_data = rbind(Subtotal_data, currentData)

  
}

if (current_orientation == "FW"){
  PrimSeqFw = as.character(PrimSeq)
  SecSeqFW = as.character(SecSeq)
}else{
  PrimSeqFw = as.character(reverseComplement(PrimSeq))
  SecSeqFW = as.character(reverseComplement(SecSeq))
}

#combine the data
Subtotal_data_names = Subtotal_data %>% 
    mutate(SangerSeqNo = clean_ab1_name_2,
           Prim_Seq = PrimSeqFw,
           Sec_Seq = SecSeqFW)

}

Total_data = rbind(Total_data, Subtotal_data_names)
}
#############################################################################################
#interpreting the outcomes & changing format
###############################################################################################

#basic interpretation
Total_data_interp = Total_data %>% mutate(Genotype = case_when(
  Seq1 == "not found" |  Seq2 == "not found" ~ "undetermined",
  Seq1 == WTseq & Seq2 == WTseq ~ "wt/wt",
  Seq1 == WTseq & Seq2 == GTseq ~ "GT/wt",
  (Seq1 != WTseq & Seq1 != GTseq & Seq1 != "not found" & Seq2 == GTseq) | (Seq2 != WTseq & Seq2 != GTseq & Seq2 != "not found" & Seq1 == GTseq) ~ "GT/mut", 
  Seq1 == GTseq & Seq2 == GTseq ~ "GT/GT",
  Seq1 == WTseq & Seq2 != WTseq & Seq2 != GTseq & Seq2 != "not found" ~ "wt/mut",
  Seq1 != GTseq & Seq1 != WTseq & Seq1 != "not found" & Seq2 != GTseq & Seq2 != WTseq & Seq2 != "not found" ~ "mut/mut",
  TRUE ~ "catch")) %>%
  mutate(Strict = strict)%>%
  #add primer info
  left_join(dna_samples, by="SangerSeqNo")

#If degenerate bases are present, match with GT seq to find if it is mut/mut or GT/mut
forbiddenletters="([RYMKSWHBVDN])"
for (i in row.names(Total_data_interp)){
  Seq2=as.character(Total_data_interp[i, "Seq2"])
  Geno=as.character(Total_data_interp[i, "Genotype"])
  if (Seq2=="not found" | Geno == "GT/GT" | Geno == "wt/wt" | Geno == "GT/wt"){
    next
  }else{
  
  Seq2_DNAString=DNAString(Seq2)
  GTseq=DNAString(as.character(Total_data_interp[i, "GTseq"]))
  if (str_detect(Seq2, forbiddenletters)==TRUE){
    degmatch = matchPattern(GTseq, Seq2_DNAString, fixed=FALSE)
    if (length(degmatch) == 1){
      if (Total_data_interp[i, "Genotype"] != "undetermined"){
      Total_data_interp[i, "Genotype"]<-"GT/mut"
      }
    }
  }
  }
}
  
  #filter away problematic primer site combinations
  no_problems = nrow(filterproblem)
  for (i in nrow(filterproblem)){
    problematic_site = filterproblem[[i, 1]]
    problematic_primer = filterproblem[[i, 2]]
    
    #get the problematic site and remove the problematic primer
    if (exists("Total_data_interp_problemfilter")==TRUE){
    Total_data_interp_problemfilter_2 = Total_data_interp_problemfilter %>% 
      mutate(Genotype = case_when(Name == problematic_site & PrimerName == problematic_primer ~ "undetermined",
                                                       TRUE ~ Genotype))
    Total_data_interp_problemfilter = Total_data_interp_problemfilter_2
    }else{
     Total_data_interp_problemfilter = Total_data_interp %>% 
      mutate(Genotype = case_when(Name == problematic_site & PrimerName == problematic_primer ~ "undetermined",
                                                       TRUE ~ Genotype))
    }
  }


  #Change the format of the data
  Data_clean = Total_data_interp_problemfilter %>% select(Name, SangerSeqNo, Genotype, Prim_Seq, Sec_Seq)
  Total_trans = spread(Data_clean, Name, Genotype)
  
  #add sample and parent info
  Total_trans_samples = merge(Total_trans, dna_samples, by="SangerSeqNo")  
  Total_trans_parents = left_join(Total_trans_samples, parentinfo, by="DNASample")  
  Total_trans_primer = left_join(Total_trans_parents, dna_orient, by="SangerSeqNo")
  

######################################################################################################### 
#then show the data per sample per locus
#determine what the outcome should be when not all sequences agree. 
#########################################################################################################

if (exists("Total_ultimate")){
rm(Total_ultimate)}

  Total_seq_rxns = Total_trans_primer %>% group_by(DNASample, Locus) %>% summarize(count_reactions = n()) %>%ungroup()

for (i in row.names(Total_seq_rxns)) {
  #initialize dataframes
  Total_consensus = Total_seq_rxns %>% dplyr::filter(row.names(Total_seq_rxns) %in% i) %>% select(DNASample, Locus)
  Total_consensus_sample = Total_consensus %>% select(DNASample)
  Total_consensus_locus = Total_consensus %>% select(Locus)
  Sample_name = as.character(Total_consensus_sample)
  Locus_name = as.character(Total_consensus_locus)
  
  for (j in row.names(sites)) {
    Site_name = as.character(sites %>% filter(row.names(sites) %in% j) %>% select(Name))
    
    #summarize so that there is not a sequence per row, but a distinct outcome per row
    Subset_working = Total_trans_primer %>%
      select(DNASample, Locus, SangerSeqNo, {{Site_name}}) %>%
      filter(DNASample == Sample_name) %>%
      filter(Locus == Locus_name) %>%
      group_by(across(c(-SangerSeqNo))) %>%
      summarize(no_observations = n()) %>%
      arrange(desc(no_observations))
    
    #remove rows where site is undetermined or possibly mutated (the latter could reflect poor sequence quality too) 
    Subset_working_2 = Subset_working %>% filter(get({{Site_name}}) != "undetermined" & get({{Site_name}}) != "wt/mut" & get({{Site_name}}) != "mut/mut" & get({{Site_name}}) != "GT/mut")
    
    #if all are undetermined or mutated make this the final outcome
    outcomes=pull(Subset_working_2, Site_name)
    outcomes_unmut=pull(Subset_working, Site_name)
    if (nrow(Subset_working_2) == 0) {
      if ("wt/mut" %in% outcomes_unmut){
        if ("GT/mut" %in% outcomes_unmut){#conflict
          Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "undetermined") 
          names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
        }else {#assume mut/mut was seq error or 1 allele had been missed
          Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "wt/mut") 
          names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
        } 
      }else if ("GT/mut" %in% outcomes_unmut){
          Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "GT/mut")
          names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
        }else if ("mut/mut" %in% outcomes_unmut){
          Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "mut/mut")
          names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
        }else{
          Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "undetermined")
          names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
        }
    #if not undetermined or mutated, the specific combination of different outcomes determines the final outcome
      }else if (nrow(Subset_working_2) > 0){
        if ("wt/wt" %in% outcomes){
          if ("GT/GT" %in% outcomes){
            Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "undetermined")
            names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
          }else if ("GT/wt" %in% outcomes){
            Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "GT/wt")  #assume 1 allele was missed in wt/wt case
            names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
          }else{
             Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "wt/wt")
            names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
          }
        }else if ("GT/GT" %in% outcomes){ #if GT/GT but not wt/wt
          if ("GT/wt" %in% outcomes){
          Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "GT/wt") #assume one allele was missed in the GT/GT case
          names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
          }else{
            Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "GT/GT")  
          names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
          }
        }else{#if GT/wt but not wt/wt and not GT/GT
            Consensus = data.frame(DNASample = {{Sample_name}}, Locus = {{Locus_name}}, temp_j = "GT/wt")
            names(Consensus)[names(Consensus) == "temp_j"] <- Site_name
        }
      }
        
    #combine data from different sites
    Total_consensus = left_join(Total_consensus, Consensus, by = c("DNASample", "Locus"))
    
  }
  
  #combine data from different samples
  if (exists("Total_ultimate") == TRUE) {
    Total_ultimate = rbind(Total_ultimate, Total_consensus)
  } else{
    Total_ultimate = Total_consensus
  }
}


#add parent info again
Total_ultimate_parents = left_join(Total_ultimate, parentinfo, by = c("DNASample"))

  
  #write everything into an excel document
work_book <- createWorkbook()
addWorksheet(work_book, "Data_detailed")
writeData(work_book, sheet = 1, Total_data_interp_problemfilter)
addWorksheet(work_book, "Data_compact")
writeData(work_book, sheet = 2, Total_trans_primer)
addWorksheet(work_book, "Data_summary")
writeData(work_book, sheet = 3, Total_ultimate_parents)

saveWorkbook(work_book, file = paste0(output_dir, "SangerSeqAnalysis.xlsx"), overwrite=TRUE)


  
  
```  

#make a figure to display the summarized data
```{r}
#process data
#note: for the future I may want to change this code to handle different numbers of sites
alldata=read.xlsx(paste0(output_dir, "SangerSeqAnalysis.xlsx"), sheet="Data_summary") %>%
  pivot_longer(cols = c(Site_1, Site_2, Site_3, Site_4, Site_5, Site_6), names_to = "Site")%>% 
  filter(Genotype!="?")%>%
  filter(!is.na(T2))%>%
  filter(Locus == "PPO_endo") %>%
  mutate(x_pos = case_when(Site == "Site_1" ~ 1,
                           Site == "Site_2" ~ 2,
                           Site == "Site_3" ~ 3,
                           Site == "Site_4" ~ 4,
                           Site == "Site_5" ~ 5,
                           Site == "Site_6" ~ 6)) %>%
  mutate(T2 = as.integer(T2))%>%
  mutate(F1 = as.integer(F1))%>%
  mutate(F2 = as.integer(F2))%>%
  arrange(T2, F1, F2) 
  
alldata_samples = alldata %>% select(DNASample) %>% unique()
alldata_samples$y_pos_sample = as.integer(row.names(alldata_samples))

alldata_ypos = left_join(alldata, alldata_samples) %>%
  left_join(fill_colour_df, by="value") 

alldata_ypos_fill_legend = alldata_ypos %>% 
  select(fill_colour, value, fill_colour_order) %>%
  unique()%>%
  arrange(fill_colour_order)


#make a dataframe with F2 group positions
alldata_ypos_F2 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
  unique()%>%
  group_by(T2, F1, F2) %>% 
  summarize(ypos_F2_stop = dplyr::last(y_pos_sample),
            count_F2=n())%>%
  mutate(ypos_F2_start = ypos_F2_stop - (count_F2-1))


#make a dataframe with F1 group positions
alldata_ypos_F1 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
    unique()%>%
  group_by(T2, F1) %>% 
  summarize(ypos_F1_stop = dplyr::last(y_pos_sample),
            count_F1=n())%>%
  mutate(ypos_F1_start = ypos_F1_stop - (count_F1-1))


#make a dataframe with T2 group positions
alldata_ypos_T2 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
    unique()%>%
  group_by(T2) %>% 
  summarize(ypos_T2_stop = dplyr::last(y_pos_sample),
            count_T2=n())%>%
  mutate(ypos_T2_start = ypos_T2_stop - (count_T2-1))

alldata_combined1=left_join(alldata_ypos, alldata_ypos_F2)
alldata_combined2=left_join(alldata_combined1, alldata_ypos_F1)
alldata_combined3=left_join(alldata_combined2, alldata_ypos_T2)

alldata_combined3$DNASample = str_replace_all(alldata_combined3$DNASample, pattern="_", replacement="-")

#to get properly formatted genotypes
tebwt = "*TEB*<sup>+/+</sup>"
tebmut= "*TEB*<sup>-/-</sup>"

no_sites = 6

#draw a stacked bar chart
plt=ggplot(data=alldata_combined3, aes(fill=fill_colour), colour="black", linewidth=line.thick)+
  #the colored rectangles
  geom_tile(aes(y=y_pos_sample, x=x_pos), height=1, width=1, linewidth=0)+
  #horizontal line at start
  geom_segment(aes(y=0.5, yend=0.5, x=0.5, xend=no_sites+0.5))+
  #horizontal lines around F2 groups
  geom_segment(aes(y=ypos_F2_stop+0.5, yend=ypos_F2_stop+0.5, x=0.5, xend=no_sites+0.5))+
  #vertical F3 lines
  #geom_segment(aes(y=y_pos_sample-0.25, yend=y_pos_sample+0.25, x=6.75, xend=6.75))+
  #F3 names next to plot
  geom_text(aes(y=y_pos_sample, x=7, label=DNASample, colour=Genotype), size=geom.text.size)+
  #vertical lines next to plot for F2 groups
  geom_segment(aes(y=ypos_F2_start-0.25, yend=ypos_F2_stop+0.25, x=7.75, xend=7.75))+
  #F2 names next to plot
  geom_text(aes(y=(ypos_F2_start+ypos_F2_stop)/2, x=8.25, label=F2, colour=Genotype), size=geom.text.size)+
  #vertical F1 group lines next to plot
  geom_segment(aes(y=ypos_F1_start-0.25, yend=ypos_F1_stop+0.25, x=8.75, xend=8.75))+
  #F1 names next to plot
  geom_text(aes(y=(ypos_F1_start+ypos_F1_stop)/2, x=9.25, label=F1), size=geom.text.size)+
  #vertical T2 group lines next to plot
  geom_segment(aes(y=ypos_T2_start-0.25, yend=ypos_T2_stop+0.25, x=9.75, xend=9.75))+
  #T2 names next to plot
  geom_text(aes(y=(ypos_T2_start+ypos_T2_stop)/2, x=10.25, label=T2), size=geom.text.size)+
  #vertical lines separating the sites
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=0.5, xend=0.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=1.5, xend=1.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=2.5, xend=2.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=3.5, xend=3.5))+  
  geom_segment(aes(y=0, yend=max(alldata_ypos$y_pos_sample)+1, x=4.5, xend=4.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=5.5, xend=5.5))+  
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=6.5, xend=6.5))+  
  #DSB position dashed horizontal line
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=4.5, xend=4.5), colour="white", linetype="dashed")+
  #legend for info below plot
  annotate("text", label="F3", y=-1, x=7, size=geom.text.size, hjust=0.5)+
  annotate("text", label="F2", y=-1, x=8.25, size=geom.text.size, hjust=0.5)+
  annotate("text", label="F1", y=-1, x=9.25, size=geom.text.size, hjust=0.5)+
  annotate("text", label="T2", y=-1, x=10.25, size=geom.text.size, hjust=0.5)+
  #rectangle color legend
  scale_fill_identity(guide="legend", breaks=alldata_ypos_fill_legend$fill_colour, labels=alldata_ypos_fill_legend$value)+
  scale_color_manual(name="", breaks=c("wt", "teb"), labels=c(tebwt, tebmut), values=c("black", "red"))+
  scale_x_continuous(breaks=c(1:no_sites), labels=c(1:no_sites))+
  labs(x="Site", y="", fill="", shape="")+
  scale_shape_identity(guide="legend", breaks=shape_df$shape, labels=shape_df$label)+
    theme(panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        text = element_text(family="Verdana", colour="black", size=theme.size),
        axis.ticks = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_text(family="Verdana", colour="black", size=theme.size),
        axis.title.x = element_text(hjust = 0.3, colour="black", size=theme.size),
        strip.background = element_blank(),
        legend.position = "right",
        legend.justification = "left",
        legend.direction = "vertical",
        legend.text = element_markdown(size=theme.size))+

  coord_cartesian(xlim=c(-0, no_sites+4.5), expand=FALSE, clip="off")

plt
ggsave(file=paste0(output_dir, "main_stackedbar.png"), plot=plt, width=14, height=30, units = "cm", dpi = "print")



```
#make a figure to display the summarized data, control data only
```{r}
#process data
alldata=read.xlsx(paste0(output_dir, "SangerSeqAnalysis.xlsx"), sheet="Data_summary") %>%
  pivot_longer(cols = c(Site_1, Site_2, Site_3, Site_4, Site_5, Site_6), names_to = "Site")%>%
  filter(Genotype=="?" | is.na(T2))%>%
  filter(DNASample != "38_1" & DNASample != "38_2")%>%
  filter(Locus == "PPO_endo") %>%
  mutate(x_pos = case_when(Site == "Site_1" ~ 1,
                           Site == "Site_2" ~ 2,
                           Site == "Site_3" ~ 3,
                           Site == "Site_4" ~ 4,
                           Site == "Site_5" ~ 5,
                           Site == "Site_6" ~ 6)) %>%
  mutate(T2 = as.integer(T2))%>%
  mutate(F1 = as.integer(F1))%>%
  mutate(F2 = as.integer(F2))%>%
  arrange(T2, F1, F2) 
  
alldata_samples = alldata %>% select(DNASample) %>% unique()
alldata_samples$y_pos_sample = as.integer(row.names(alldata_samples))

alldata_ypos = left_join(alldata, alldata_samples) %>%
  left_join(fill_colour_df, by="value") 

alldata_ypos_fill_legend = alldata_ypos %>% 
  select(fill_colour, value, fill_colour_order) %>%
  unique()%>%
  arrange(fill_colour_order)


#make a dataframe with F2 group positions
alldata_ypos_F2 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
  unique()%>%
  group_by(T2, F1, F2) %>% 
  summarize(ypos_F2_stop = dplyr::last(y_pos_sample),
            count_F2=n())%>%
  mutate(ypos_F2_start = ypos_F2_stop - (count_F2-1))


#make a dataframe with F1 group positions
alldata_ypos_F1 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
    unique()%>%
  group_by(T2, F1) %>% 
  summarize(ypos_F1_stop = dplyr::last(y_pos_sample),
            count_F1=n())%>%
  mutate(ypos_F1_start = ypos_F1_stop - (count_F1-1))


#make a dataframe with T2 group positions
alldata_ypos_T2 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
    unique()%>%
  group_by(T2) %>% 
  summarize(ypos_T2_stop = dplyr::last(y_pos_sample),
            count_T2=n())%>%
  mutate(ypos_T2_start = ypos_T2_stop - (count_T2-1))

alldata_combined1=left_join(alldata_ypos, alldata_ypos_F2)
alldata_combined2=left_join(alldata_combined1, alldata_ypos_F1)
alldata_combined3=left_join(alldata_combined2, alldata_ypos_T2)

tebwt=expression(italic(TEB)^"+/+")
tebmut=expression(italic(TEB)^"-/-")

no_sites = 6

#draw a stacked bar chart
plt=ggplot(data=alldata_combined3, aes(fill=fill_colour), colour="black", linewidth=line.thick)+
  #the colored rectangles
  geom_tile(aes(y=y_pos_sample, x=x_pos), height=1, width=1, linewidth=0)+
  #horizontal line at start
  geom_segment(aes(y=0.5, yend=0.5, x=0.5, xend=no_sites+0.5))+
  #horizontal line around F1 groups
  geom_segment(aes(y=ypos_F1_stop+0.5, yend=ypos_F1_stop+0.5, x=0.5, xend=no_sites+0.5))+
  #vertical lines on the right for F2 names
  #geom_segment(aes(y=y_pos_sample-0.25, yend=y_pos_sample+0.25, x=7, xend=7))+
  #F2 names on the right
  geom_text(aes(y=y_pos_sample, x=7, label=DNASample), size=geom.text.size)+
  #vertical lines on the right for F1 groups
  geom_segment(aes(y=ypos_F1_start-0.25, yend=ypos_F1_stop+0.25, x=7.75, xend=7.75))+
  #F1 names on the right
  geom_text(aes(y=(ypos_F1_start+ypos_F1_stop)/2, x=8.25, label=F1), size=geom.text.size)+
  #vertical lines on the right for T2 groups
  geom_segment(aes(y=ypos_T2_start-0.25, yend=ypos_T2_stop+0.25, x=8.75, xend=8.75))+
  #T2 names on the right
  geom_text(aes(y=(ypos_T2_start+ypos_T2_stop)/2, x=9.25, label=T2), size=geom.text.size)+
  #horizontal lines separating the sites
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=0.5, xend=0.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=1.5, xend=1.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=2.5, xend=2.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=3.5, xend=3.5))+  
  geom_segment(aes(y=0, yend=max(alldata_ypos$y_pos_sample)+1, x=4.5, xend=4.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=5.5, xend=5.5))+  
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=6.5, xend=6.5))+  
  #DSB position dashed horizontal line
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=4.5, xend=4.5), colour="white", linetype="dashed")+
  #legend for info below plot
  annotate("text", label="F2", y=-1, x=7, size=geom.text.size, hjust=0.5)+
  annotate("text", label="F1", y=-1, x=8.25, size=geom.text.size, hjust=0.5)+
  annotate("text", label="T2", y=-1, x=9.25, size=geom.text.size, hjust=0.5)+
  #rectangle color legend
  #rectangle color legend
  scale_fill_identity(guide="legend", breaks=alldata_ypos_fill_legend$fill_colour, labels=alldata_ypos_fill_legend$value)+
  #scale_color_manual(name="F2 bkgd:", breaks=c("wt", "teb"), labels=c(tebwt, tebmut), values=c("black", "red"))+
  scale_x_continuous(breaks=c(1:no_sites), labels=c(1:no_sites))+
  labs(x="Site", y="", fill="", shape="")+
    theme(panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        text = element_text(family="Verdana", colour="black", size=theme.size),
        axis.ticks = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_text(family="Verdana", colour="black", size=theme.size),
        axis.title.x = element_text(hjust = 0.3, colour="black", size=theme.size),
        strip.background = element_blank(),
        legend.position = "right",
        legend.justification = "left",
        legend.direction = "vertical",
        legend.text = element_markdown(size=theme.size))+
  coord_cartesian(xlim=c(-0, no_sites+3.5), expand=FALSE, clip="off")

plt
ggsave(file=paste0(output_dir, "main_stackedbar_sup.png"), width=15, height=15, units = "cm", dpi = "print")



```
#make a figure to display the summarized data of the ectopic GT events
```{r}
#process data
#note: for the future I may want to change this code to handle different numbers of sites
alldata=read.xlsx(paste0(output_dir, "SangerSeqAnalysis.xlsx"), sheet="Data_summary") %>%
  pivot_longer(cols = c(Site_1, Site_2, Site_3, Site_4, Site_5, Site_6), names_to = "Site")%>% 
  filter(Genotype!="?")%>%
  filter(!is.na(T2))%>%
  filter(Locus == "PPO_exo") %>%
  mutate(x_pos = case_when(Site == "Site_1" ~ 1,
                           Site == "Site_2" ~ 2,
                           Site == "Site_3" ~ 3,
                           Site == "Site_4" ~ 4,
                           Site == "Site_5" ~ 5,
                           Site == "Site_6" ~ 6)) %>%
  mutate(T2 = as.integer(T2))%>%
  mutate(F1 = as.integer(F1))%>%
  mutate(F2 = as.integer(F2))%>%
  arrange(T2, F1, F2) 
  
alldata_samples = alldata %>% select(DNASample) %>% unique()
alldata_samples$y_pos_sample = as.integer(row.names(alldata_samples))

alldata_ypos = left_join(alldata, alldata_samples) %>%
  left_join(fill_colour_df, by="value") 

alldata_ypos_fill_legend = alldata_ypos %>% 
  select(fill_colour, value, fill_colour_order) %>%
  unique()%>%
  arrange(fill_colour_order)


#make a dataframe with F2 group positions
alldata_ypos_F2 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
  unique()%>%
  group_by(T2, F1, F2) %>% 
  summarize(ypos_F2_stop = dplyr::last(y_pos_sample),
            count_F2=n())%>%
  mutate(ypos_F2_start = ypos_F2_stop - (count_F2-1))


#make a dataframe with F1 group positions
alldata_ypos_F1 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
    unique()%>%
  group_by(T2, F1) %>% 
  summarize(ypos_F1_stop = dplyr::last(y_pos_sample),
            count_F1=n())%>%
  mutate(ypos_F1_start = ypos_F1_stop - (count_F1-1))


#make a dataframe with T2 group positions
alldata_ypos_T2 = alldata_ypos %>% 
  select(T2, F1, F2, y_pos_sample) %>%
    unique()%>%
  group_by(T2) %>% 
  summarize(ypos_T2_stop = dplyr::last(y_pos_sample),
            count_T2=n())%>%
  mutate(ypos_T2_start = ypos_T2_stop - (count_T2-1))

alldata_combined1=left_join(alldata_ypos, alldata_ypos_F2)
alldata_combined2=left_join(alldata_combined1, alldata_ypos_F1)
alldata_combined3=left_join(alldata_combined2, alldata_ypos_T2)

alldata_combined3$DNASample = str_replace_all(alldata_combined3$DNASample, pattern="_", replacement="-")

#to get properly formatted genotypes
tebwt = "*TEB*<sup>+/+</sup>"
tebmut= "*TEB*<sup>-/-</sup>"

no_sites = 6

#draw a stacked bar chart
plt=ggplot(data=alldata_combined3, aes(fill=fill_colour), colour="black", linewidth=line.thick)+
  #the colored rectangles
  geom_tile(aes(y=y_pos_sample, x=x_pos), height=1, width=1, linewidth=0)+
  #horizontal line at start
  geom_segment(aes(y=0.5, yend=0.5, x=0.5, xend=no_sites+0.5))+
  #horizontal lines around F2 groups
  geom_segment(aes(y=ypos_F2_stop+0.5, yend=ypos_F2_stop+0.5, x=0.5, xend=no_sites+0.5))+
  #vertical F3 lines
  #geom_segment(aes(y=y_pos_sample-0.25, yend=y_pos_sample+0.25, x=7, xend=7))+
  #F3 names next to plot
  geom_text(aes(y=y_pos_sample, x=7, label=DNASample, colour=Genotype), size=geom.text.size)+
  #vertical lines next to plot for F2 groups
  geom_segment(aes(y=ypos_F2_start-0.25, yend=ypos_F2_stop+0.25, x=7.75, xend=7.75))+
  #F2 names next to plot
  geom_text(aes(y=(ypos_F2_start+ypos_F2_stop)/2, x=8.25, label=F2, colour=Genotype), size=geom.text.size)+
  #vertical F1 group lines next to plot
  geom_segment(aes(y=ypos_F1_start-0.25, yend=ypos_F1_stop+0.25, x=8.75, xend=8.75))+
  #F1 names next to plot
  geom_text(aes(y=(ypos_F1_start+ypos_F1_stop)/2, x=9.25, label=F1), size=geom.text.size)+
  #vertical T2 group lines next to plot
  geom_segment(aes(y=ypos_T2_start-0.25, yend=ypos_T2_stop+0.25, x=9.75, xend=9.75))+
  #T2 names next to plot
  geom_text(aes(y=(ypos_T2_start+ypos_T2_stop)/2, x=10.25, label=T2), size=geom.text.size)+
  #vertical lines separating the sites
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=0.5, xend=0.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=1.5, xend=1.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=2.5, xend=2.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=3.5, xend=3.5))+  
  geom_segment(aes(y=0, yend=max(alldata_ypos$y_pos_sample)+1, x=4.5, xend=4.5))+
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=5.5, xend=5.5))+  
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=6.5, xend=6.5))+  
  #DSB position dashed horizontal line
  geom_segment(aes(y=0.5, yend=max(alldata_ypos$y_pos_sample)+0.5, x=4.5, xend=4.5), colour="white", linetype="dashed")+
  #legend for info below plot
  annotate("text", label="F3", y=-1, x=7, size=geom.text.size, hjust=0.5)+
  annotate("text", label="F2", y=-1, x=8.25, size=geom.text.size, hjust=0.5)+
  annotate("text", label="F1", y=-1, x=9.25, size=geom.text.size, hjust=0.5)+
  annotate("text", label="T2", y=-1, x=10.25, size=geom.text.size, hjust=0.5)+
  #rectangle color legend
  scale_fill_identity(guide="legend", breaks=alldata_ypos_fill_legend$fill_colour, labels=alldata_ypos_fill_legend$value)+
  scale_color_manual(name="", breaks=c("wt", "teb"), labels=c(tebwt, tebmut), values=c("black", "red"))+
  scale_x_continuous(breaks=c(1:no_sites), labels=c(1:no_sites))+
  labs(x="Site", y="", fill="", shape="")+
  scale_shape_identity(guide="legend", breaks=shape_df$shape, labels=shape_df$label)+
    theme(panel.border = element_blank(), 
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        text = element_text(family="Verdana", colour="black", size=theme.size),
        axis.ticks = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_text(family="Verdana", colour="black", size=theme.size),
        axis.title.x = element_text(hjust = 0.3, colour="black", size=theme.size),
        strip.background = element_blank(),
        legend.position = "right",
        legend.justification = "left",
        legend.direction = "vertical",
        legend.text = element_markdown(size=theme.size))+

  coord_cartesian(xlim=c(-0, no_sites+4.5), expand=FALSE, clip="off")

plt
ggsave(file=paste0(output_dir, "ectopicGT_stackedbar.png"), plot=plt, width=14, height=6, units = "cm", dpi = "print")



```

#gene conversion tract length figure
```{r}
#start with site 4. check whether GT/wt or GT/GT.
#then check how far the same color continues in one direction, and in the other direction.
#do not use rows with wt/mut, wt/wt, gt/mut, mut/mut, undetermined
#display the lengths per F2, per genotype.

alldata=read.xlsx(paste0(output_dir, "SangerSeqAnalysis.xlsx"), sheet="Data_summary") %>%
  #first get rid of samples where some information is unclear
  filter(Site_1 != "undetermined" & Site_2 != "undetermined" & Site_3 != "undetermined" & Site_4 != "undetermined" & Site_5 != "undetermined" & Site_6 != "undetermined" )%>%
  #then get rid of the ones without ATG at codon 426
  filter(Site_4!="wt/wt")%>%
  #only look at endogenous locus
  filter(Locus=="PPO_endo")%>%
  #simplify the calling
  mutate(Site_1 = case_when(Site_1 == "GT/GT" | Site_1 == "GT/wt" | Site_1== "GT/mut" ~ TRUE,
                            TRUE ~ FALSE))%>%
  mutate(Site_2 = case_when(Site_2 == "GT/GT" | Site_2 == "GT/wt" | Site_2== "GT/mut" ~ TRUE,
                            TRUE ~ FALSE))%>%
  mutate(Site_3 = case_when(Site_3 == "GT/GT" | Site_3 == "GT/wt" | Site_3== "GT/mut" ~ TRUE,
                            TRUE ~ FALSE))%>%
  mutate(Site_4 = case_when(Site_4 == "GT/GT" | Site_4 == "GT/wt" | Site_4== "GT/mut" ~ TRUE,
                            TRUE ~ FALSE))%>%
  mutate(Site_5 = case_when(Site_5 == "GT/GT" | Site_5 == "GT/wt" | Site_5== "GT/mut" ~ TRUE,
                            TRUE ~ FALSE))%>%
  mutate(Site_6 = case_when(Site_6 == "GT/GT" | Site_6 == "GT/wt" | Site_6== "GT/mut" ~ TRUE,
                            TRUE ~ FALSE))%>%
  mutate(uplen = 99999999,
         downlen = 99999999)

for (i in row.names(alldata)){
  number=as.integer(i)
  site_1_gt = alldata$Site_1[number]
  site_2_gt = alldata$Site_2[number]
  site_3_gt = alldata$Site_3[number]
  site_5_gt = alldata$Site_5[number]
  site_6_gt = alldata$Site_6[number]
  
  if (site_3_gt == TRUE){
    if (site_2_gt == TRUE){
      if (site_1_gt == TRUE){
        alldata$uplen[number]=693
      }else{
        alldata$uplen[number]=544
      }
    }else{
      alldata$uplen[number]=429
    }
  }else{
    alldata$uplen[number]=5
  }
  if (site_5_gt == TRUE){
    if (site_6_gt == TRUE){
      alldata$downlen[number] = 118
    }else{
      alldata$downlen[number] = 58
    }
  }else{
    alldata$downlen[number] = 0
  }
}

data_figure=alldata %>%
  filter(Genotype!="?")%>%
  mutate(uplen=(uplen*-1))%>%
  rowwise()%>%
  mutate(parent = paste0(T2, F1, F2))%>%
  ungroup()%>%
  group_by(parent, Genotype)%>%
  summarize(mean_uplen = mean(uplen),
            sd_uplen = sd(uplen),
            mean_downlen = mean(downlen),
            sd_downlen = sd(downlen),
            n =n())%>%
  rowwise()%>%
  mutate(SE_uplen = sd_uplen/sqrt(n),
         SE_downlen = sd_downlen/sqrt(n))%>%
  replace_na(list(SE_uplen = 0, SE_downlen = 0))%>%
  mutate(SE_up = mean_uplen - SE_uplen,
         SE_down = mean_downlen + SE_downlen)%>%
  mutate(totallen = mean_downlen + (mean_uplen*-1))%>%
  ungroup()%>%
  group_by(Genotype)%>%
  arrange(totallen, .by_group=TRUE)

tebwt = "*TEB*<sup>+/+</sup>"
tebmut= "*TEB*<sup>-/-</sup>"

data_figure_teb = data_figure %>%
  filter(Genotype== "teb")%>%
  mutate(Genotype_nicer = tebmut)
data_figure_wt = data_figure %>%
  filter(Genotype== "wt")%>%
  mutate(Genotype_nicer = tebwt)

data_figure_teb = rownames_to_column(data_figure_teb, "rown")%>%
  rowwise()%>%
  mutate(rown = as.integer(rown))
data_figure_wt = rownames_to_column(data_figure_wt, "rown")%>%
  rowwise()%>%
  mutate(rown = as.integer(rown)+3)



data_figure_combined = rbind(data_figure_teb, data_figure_wt)

data_figure_combined$Genotype_nicer = factor(data_figure_combined$Genotype_nicer, levels=c(tebwt, tebmut))


plt = ggplot(data=data_figure_combined, aes(y=rown))+
  geom_col(aes(x=mean_downlen), fill="grey90", col="black", orientation = "y", linewidth=line.thick)+
  geom_col(aes(x=mean_uplen), fill="grey90", col="black", orientation = "y", linewidth=line.thick)+
  labs(x="Mean gene conversion tract length (bp from DSB)", y="F3s grouped by F2 parent")+
    theme(
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_rect(fill="white", colour="black", linewidth=line.thick*2), 
        text = element_text(family="Verdana", colour="black", size=theme.size),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_text(family="Verdana", colour="black", size=theme.size),
        axis.title.x = element_text(colour="black", size=theme.size),
        strip.text = element_markdown(size=theme.size),
        strip.background = element_blank(),
        legend.position = "right",
        legend.justification = "left",
        legend.direction = "vertical",
        legend.text = element_markdown(size=theme.size))+
  geom_vline(xintercept=-693, lty="dotted", col="red", linewidth=line.thick)+
  geom_vline(xintercept=118, lty="dotted", col="red", linewidth=line.thick)+
  geom_errorbar(aes(xmin=SE_up, xmax = mean_uplen), linewidth=line.thick)+
  geom_errorbar(aes(xmin=mean_downlen, xmax = SE_down), linewidth=line.thick)+
  scale_x_continuous(breaks=seq(from=-700, to=100, by=100))+
  facet_grid(cols=vars(Genotype_nicer))


plt
ggsave(file=paste0(output_dir, "conversion_tract.png"), width=16, height=8, units = "cm", dpi = "print")




#statistical test to determine if teb and wt have different average tract length
wilcox.exact(data_figure_combined$totallen ~ data_figure_combined$Genotype, paired=FALSE, alternative="two.sided")
median(data_figure_teb$totallen)
median(data_figure_wt$totallen)

```



